src/net.cpp

```diff
@@ -1152,6 +1152,37 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     RandAddEvent((uint32_t)id);
 }
 
+bool CConnman::AddConnection(const std::string& address, const ConnectionType conn_type)
+{
+    assert(conn_type == ConnectionType::OUTBOUND_FULL_RELAY || conn_type == ConnectionType::BLOCK_RELAY);
+
+    int outbound_full_relay = 0;
+    int outbound_block_relay = 0;
+
+    {
+        // count existing connections for each type
+        LOCK(cs_vNodes);
+        for (const CNode* pnode : vNodes) {
+            if (pnode->m_conn_type == ConnectionType::OUTBOUND_FULL_RELAY) {
+                ++outbound_full_relay;
+            } else if (pnode->m_conn_type == ConnectionType::BLOCK_RELAY) {
+                ++outbound_block_relay;
+            }
+        }
+    }
+
+    // max connections of specified type already exist
+    if (conn_type == ConnectionType::OUTBOUND_FULL_RELAY && outbound_full_relay >= m_max_outbound_full_relay) return false;
+    if (conn_type == ConnectionType::BLOCK_RELAY && outbound_block_relay >= m_max_outbound_block_relay) return false;
+
+    // max total outbound connections already exist
+    CSemaphoreGrant grant(*semOutbound, true);
```

Comment:
I don't understand exactly what `CSemaphoreGrant` is enforcing given that:
* `semOutbound` is initialized with a value of `std::min(m_max_outbound, nMaxConnections)` (L2510 in `src/net.cpp`)
* thread "addconnection" is accessing this semaphore L1179 in `src/net.cpp` with `fTry=true` thus calling `CSemaphoreGrant::TryAcquire`, decrementing the semaphore by one (L281 in `src/sync.h`) and returning a successful grant taking
* thread "addconnection is calling `OpenNetworkConnection` which itself calls `CSemaphoreGrant::MoveTo()` (L2205 in `src/net.cpp`) if connection opening is successful. `MoveTo()` calls `CSemaphoreGrant::Release()` then `CSemaphore::post()`. This last function _increments_ back the semaphore.
* a thread `opencon` is accessing the semaphore in `ThreadOpenConnections` (L1918) and indirectly calls `CSemaphore::wait`, but the conditional predicate doesn't matter as `addconnection` thread incremented back the semaphore. Thus sooner or latter `wait()` will return and a new outbound should be established.

`CSemaphoreGrant`/`CSemaphore`/`CConnman::semOutbound` are poorly documented so it's hard to understand what exactly it aims to achieve. So we might theoritcally bypass `m_max_outbound` but I wouldn't surprised you see nothing in practice as block-relay-only/full-relay are checked on their own and other outbound are short-lived ?

Reply:
`a.MoveTo(b)` calls `b.Release()` -- but `b` in this case is a brand new `CSemaphoreGrant` that was default initialized, so `b.fHaveGrant` is false and `b.Release()` is a no-op, and `b.sem->post()` is not called.

Reply:
@ariard I agree that `CSemaphoreGrant` is far-from-ideal, and that its strange / redundant that we also have the individual counts of our different types. However, I'm not understanding exactly what you're trying to suggest here. Are you saying that the overall design of this mechanism is confusing, and the guarantees minimal? Or are you saying that this proposed implementation is problematic?

Reply:
Thanks @ajtowns  for the code reading correction! Okay so the semaphore should accomplish its design of preventing to open too much outbound connections.

@amitiuttarwar yes I was saying the overall design of this mechanism is confusing thus making it hard to understand the proposed implementation, but not an issue introduced by the present PR. `CSemaphoreGrant` would gain to be documented better but not proposing myself for this :see_no_evil:

Reply:
haha, I fully agree 😛

going to resolve this conversation to help me keep track of outstanding issues on this PR